File: Sources/CoMaintainerCore/PRRef.swift
import Foundation

/// A PR reference parsed from a single-PR text field.
///
/// The field accepts a bare number (`337`, `#337`), a full GitHub PR URL
/// (`https://github.com/owner/repo/pull/337`, with any trailing path/query), or the
/// `owner/repo#337` shorthand. When the input names a repo (URL or shorthand) it's
/// checked against the configured target repo, so a link to the wrong project is
/// rejected instead of silently reviewing the wrong PR. Shared verbatim with the
/// Linux front-end (see `prref.py`).
public struct PRRef: Equatable {
    /// The extracted PR number, or nil when the input contains a usable one.
    public let number: Int?
    /// True only when the input named a repo that does NOT match the target. A bare
    /// number names no repo, so it never mismatches.
    public let repoMismatch: Bool

    public init(number: Int?, repoMismatch: Bool) {
        self.number = number
        self.repoMismatch = repoMismatch
    }

    /// A usable reference: a number was found and any named repo matched.
    public var isValid: Bool { number != nil && !repoMismatch }
    /// The bare number for prompt injection ("" when none).
    public var numberString: String { number.map(String.init) ?? "" }

    /// Parse `raw` against the expected `owner`/`repo` (compared case-insensitively).
    public static func parse(_ raw: String, owner: String, repo: String) -> PRRef {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return PRRef(number: nil, repoMismatch: false) }

        // A GitHub PR URL anywhere in the input: github.com/OWNER/REPO/pull/N.
        // [0-9], not \d: ICU's \d matches non-ASCII digits, which Int() then rejects —
        // and the Python port matches ASCII only, so the two sides would disagree.
        if let g = firstMatch(s, #"(?:https?://)?(?:www\.)?github\.com/([\w.-]+)/([\w.-]+)/pull/([0-9]+)"#) {
            return named(owner: g[1], repo: g[2], number: g[3], wantOwner: owner, wantRepo: repo)
        }
        // The OWNER/REPO#N shorthand (the whole string).
        if let g = firstMatch(s, #"^([\w.-]+)/([\w.-]+)#([0-9]+)$"#) {
            return named(owner: g[1], repo: g[2], number: g[3], wantOwner: owner, wantRepo: repo)
        }
        // A bare number, with an optional leading '#'. ASCII digits only — Int() alone
        // also accepts "+337", which the Python port rejects.
        let bare = s.hasPrefix("#") ? String(s.dropFirst()) : s
        if !bare.isEmpty, bare.allSatisfy({ $0.isASCII && $0.isNumber }),
           let n = Int(bare), n > 0 { return PRRef(number: n, repoMismatch: false) }

        return PRRef(number: nil, repoMismatch: false)
    }

    private static func named(owner: String, repo: String, number: String,
                              wantOwner: String, wantRepo: String) -> PRRef {
        let n = Int(number).flatMap { $0 > 0 ? $0 : nil }
        let matches = owner.caseInsensitiveCompare(wantOwner) == .orderedSame
            && repo.caseInsensitiveCompare(wantRepo) == .orderedSame
        return PRRef(number: n, repoMismatch: !matches)
    }

    /// The capture groups (index 0 = whole match, 1…n = groups) of the first match
    /// of `pattern` in `s`, or nil when it doesn't match.
    private static func firstMatch(_ s: String, _ pattern: String) -> [String]? {
        guard let re = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else { return nil }
        let full = NSRange(s.startIndex..., in: s)
        guard let m = re.firstMatch(in: s, options: [], range: full) else { return nil }
        return (0..<m.numberOfRanges).map { i in
            Range(m.range(at: i), in: s).map { String(s[$0]) } ?? ""
        }
    }
}

File: packages/diplomat-runtime/diplomat_runtime/prref.py
"""Parse a PR or issue reference from a wizard's single-item text field.

Mirrors ``PRRef.swift`` verbatim: accepts a bare number (``337`` / ``#337``), a
full GitHub URL (``https://github.com/owner/repo/pull/337`` with any trailing
path/query), or the ``owner/repo#337`` shorthand. When the input names a repo it's
checked against the configured target repo, so a link to the wrong project is
rejected instead of silently reviewing the wrong PR. ``kind`` picks the URL path
segment a pasted link must carry — a PR link and an issue link differ by that
segment alone, and accepting either everywhere would let an issue URL through the
Review wizard as a PR number.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# PR numbers are ASCII digits only ([0-9], not \d): Python's \d / str.isdigit()
# also match non-ASCII digits like "٣٣٧", which Swift's Int(_:) rejects.
# github.com/OWNER/REPO/<kind>/N — scheme/www optional, trailing path/query allowed.
_URL = {
    kind: re.compile(
        rf"(?:https?://)?(?:www\.)?github\.com/([\w.-]+)/([\w.-]+)/{kind}/([0-9]+)",
        re.IGNORECASE,
    )
    for kind in ("pull", "issues")
}
# OWNER/REPO#N shorthand (whole string).
_SHORTHAND = re.compile(r"^([\w.-]+)/([\w.-]+)#([0-9]+)$")
# A bare PR number (after any leading '#' is stripped).
_BARE_NUMBER = re.compile(r"^[0-9]+$")

# Swift's ``Int`` is 64-bit, so ``Int("9223372036854775808")`` (2^63, one past
# Int64.max) overflows to nil and PRRef.swift rejects it. Python's ``int()`` is
# unbounded and would accept it, diverging the two front-ends. Clamp here so an
# over-large PR number is rejected on both (PRRef.swift parity).
_INT64_MAX = 9223372036854775807


def _int64(digits: str) -> int | None:
    """The value of an all-ASCII-digit string as Swift's ``Int(_:)`` would see it:
    the integer, or ``None`` when it overflows the 64-bit range."""
    n = int(digits)  # digits is [0-9]+, so this never raises
    return n if n <= _INT64_MAX else None


@dataclass(frozen=True)
class PRRef:
    number: int | None
    repo_mismatch: bool

    @property
    def is_valid(self) -> bool:
        """A usable reference: a number was found and any named repo matched."""
        return self.number is not None and not self.repo_mismatch

    @property
    def number_string(self) -> str:
        """The bare number for prompt injection ("" when none)."""
        return str(self.number) if self.number is not None else ""


def parse_pr_ref(raw: str, owner: str, repo: str, kind: str = "pull") -> PRRef:
    """Parse ``raw`` against the expected ``owner``/``repo`` (case-insensitive).
    ``kind`` is ``"pull"`` for a PR field or ``"issues"`` for an issue one."""
    s = raw.strip()
    if not s:
        return PRRef(None, False)

    m = _URL[kind].search(s) or _SHORTHAND.match(s)
    if m:
        o, r, n = m.group(1), m.group(2), _int64(m.group(3))
        matches = o.lower() == owner.lower() and r.lower() == repo.lower()
        return PRRef(n if (n is not None and n > 0) else None, not matches)

    bare = s[1:] if s.startswith("#") else s
    if _BARE_NUMBER.match(bare):
        n = _int64(bare)
        if n is not None and n > 0:
            return PRRef(n, False)

    return PRRef(None, False)
